pygpu/array_tools.py

```diff
@@ -0,0 +1,47 @@
+from string import Template
+from .gpuarray import GpuArray, GpuKernel
+
+
+def _generate_kernel(ctx, cols, upper=True):
+    tmpl = Template("""
+    KERNEL void extract_tri(ga_float *a, ga_uint N) {
+        unsigned int idx = blockIdx.y*blockDim.x*gridDim.x+
+                           blockIdx.x*blockDim.x+threadIdx.x;
+        unsigned int ix = idx/${cols};
+        unsigned int iy = idx%${cols};
+        if (idx < N) {
+            if (ix ${le} iy)
+                a[idx] = 0.0;
+        }
+    }
+    """)
+    if upper:
+        le = '>'
+    else:
+        le = '<'
+    src = tmpl.substitute(cols=cols, le=le)
+    spec = [GpuArray, 'uint32']
+    k = GpuKernel(src, "extract_tri", spec, context=ctx)
+    return k
+
+
+def triu(A, ctx, inplace=True):
```

Comment:
As `A` is a GPU array, `triu` and `tril` can deduce `ctx` arg from `A`, that is `A.context`. So, rewrite both functions so that they take only 2 args (A and inplace).